split register_aten_ops.cpp into shards

anderspapitto opened the pull request:
after an analogous breakup of VariableType.cpp, the generated
register_aten_ops.cpp is now the slowest-to-compile file in a typical
incremental rebuild by a wide margin. Therefore, give it the same
treatment - the generated code is split across several files to allow
parallel compilation.

Note that the existing code takes some care to arrange that overloads
of the same op name are given in a particular order. This diff
preserves that behavior, by treating all overloads of the same name as
a single indivisible unit, and sharding based on these groups rather
than on individual constructors.

ezyang commented on tools/jit/templates/register_aten_ops.cpp:
```diff
@@ -35,7 +44,7 @@ using at::DeviceGuard;
 namespace {
 
-int deviceForInputs(Stack & stack, size_t N) {
+static int deviceForInputs(Stack & stack, size_t N) {
```
(A more explicit way to say, "I just want this to inline always" is to mark the function as inline)

anderspapitto replied:
will switch to inline

apaszke replied:
FWIW `inline` doesn't actually affect the function inlining in any way (e.g. inline functions can be recursive!). It only affects the linkage of this function - linker will be made to deduplicate different instantiations.

ezyang commented on tools/jit/templates/register_aten_ops.cpp:
```diff
@@ -24,6 +24,15 @@
 
 // ${generated_comment}
 
+// NOTE: This file is generated in a sharded fashion to speed up
```
The convention in the codebase is to give a note a name, like

```
// Note [Sharded file]
// This file is blah blah blah
```

And then refer to the note using `See Note [Sharded file]`. This allows devs to easily grep for all sites of a note reference.

anderspapitto replied:
will do

ezyang commented on tools/jit/gen_jit_dispatch.py:
```diff
@@ -274,16 +270,27 @@ def declkey(decl):
+
+    # NOTE: see the comment at the top of the register_aten_ops.cpp
+    # template regarding sharding of the generated files.
+    num_shards = 3
```
This should say where you have to edit if the number of shards changes

anderspapitto replied:
will do

ezyang commented on tools/jit/gen_jit_dispatch.py:
```diff
@@ -274,16 +270,27 @@ def declkey(decl):
+
+    # ops are assigned arbitrarily but stably to a file based on hash
+    for group in jit_decl_groups:
+        x = sum(ord(c) for c in group[0]['name']) % num_shards
```
@zdevito, I know you suggested this sharding strategy, but wouldn't it be better to use an actual hash function here? Like, it's not like Python makes it hard to do so. (I agree that this particular use site doesn't matter how we do it; it's more of a "is this code going to get copy pasted" concern)

anderspapitto replied:
well this already is a copy-paste of the other site :)

I could switch to just calling hash(). It looks like it's deterministic across interpreter startups (I wasn't sure about this, but I've just tested it). It may vary between python versions, or maybe even between machines or environments? Not sure about this, or if it matters even if true

apaszke replied:
No, `hash` is not guaranteed to be consistent, and in fact won't be in many situations. We need to have our own custom function

ezyang commented on tools/jit/gen_jit_dispatch.py:
```diff
@@ -274,16 +270,27 @@ def declkey(decl):
+        env = {
+            'constructors': shard,
+        }
+        write(out, 'register_aten_ops-%d.cpp' % i, REGISTER_ATEN_OPS_CPP, env)
```
This is not portable to Python 3. Use 'register_aten_ops-{}.cpp'.format(i) instead

ezyang reviewed (approved):
I feel a little nervous about the hard-coding of sharded file names, but I guess this is OK for now.

anderspapitto commented:
yeah, hard-coding the sharding split is a bit ugly but

- it won't change often
- it's way simpler than threading the value through to cmake/buck/python from some single source of truth
- if you get it wrong you get super obvious errors

apaszke commented on tools/setup_helpers/generate_code.py:
```diff
@@ -44,7 +44,9 @@ def all_generator_source():
-    'torch/csrc/jit/generated/register_aten_ops.cpp',
+    'torch/csrc/jit/generated/register_aten_ops-0.cpp',
+    'torch/csrc/jit/generated/register_aten_ops-1.cpp',
+    'torch/csrc/jit/generated/register_aten_ops-2.cpp',
```
Uh Python has a great library called `glob`. Can we just include all of `torch/csrc/jit/generated/register_aten_ops-*.cpp`?

apaszke reviewed (commented):
Can we please switch the hyphen to an underscore? This file is very handy when editing JIT sources, and typing in those characters will be super annoying.

anderspapitto committed:
split register_aten_ops.cpp into shards (#12615)

Summary:
after an analogous breakup of VariableType.cpp, the generated
register_aten_ops.cpp is now the slowest-to-compile file in a typical
incremental rebuild by a wide margin. Therefore, give it the same
treatment - the generated code is split across several files to allow
parallel compilation.

Note that the existing code takes some care to arrange that overloads
of the same op name are given in a particular order. This diff
preserves that behavior, by treating all overloads of the same name as
a single indivisible unit, and sharding based on these groups rather
than on individual constructors.
Pull Request resolved: https://github.com/pytorch/pytorch/pull/12615

Differential Revision: D10367363

fbshipit-source-id: 82029f285f72817b18d4ad5261a7b408d16125d1
